src/block/Composter.php

```diff
@@ -0,0 +1,139 @@
+ * |_|   \___/ \___|_|\_\___|\__|_|  |_|_|_| |_|\___|     |_|  |_|_|
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * @author PocketMine Team
+ * @link http://www.pocketmine.net/
+ *
+ *
+ */
+
+declare(strict_types=1);
+
+namespace pocketmine\block;
+
+use pocketmine\data\runtime\RuntimeDataDescriber;
+use pocketmine\event\block\ComposterEmptyEvent;
+use pocketmine\event\block\ComposterFillEvent;
+use pocketmine\event\block\ComposterReadyEvent;
+use pocketmine\item\Item;
+use pocketmine\item\VanillaItems;
+use pocketmine\math\AxisAlignedBB;
+use pocketmine\math\Facing;
+use pocketmine\math\Vector3;
+use pocketmine\player\Player;
+use pocketmine\world\particle\CropGrowthEmitterParticle;
+use pocketmine\world\sound\ComposterEmptySound;
+use pocketmine\world\sound\ComposterFillSound;
+use pocketmine\world\sound\ComposterFillSuccessSound;
+use pocketmine\world\sound\ComposterReadySound;
+
+class Composter extends Transparent{
+	public const MIN_COMPOST_LAYERS = 0;
+	public const MAX_COMPOST_LAYERS = 8;
+
+	protected int $layers = 0;
+
+	protected function describeBlockOnlyState(RuntimeDataDescriber $w) : void{
+		$w->boundedIntAuto(0, self::MAX_COMPOST_LAYERS, $this->layers);
+	}
+	
+	protected function recalculateCollisionBoxes() : array{
+		foreach(Facing::HORIZONTAL as $f){
+			$result[] = AxisAlignedBB::one()->trim($f, 14 / 16);
+		}
+		return $result;
+	}
+
+	public function getCompostLayers() : int{ return $this->layers; }
+
+	/** @return $this */
+	public function setCompostLayers(int $layers) : self{
+		if($layers < 0 || $layers > self::MAX_COMPOST_LAYERS){
+			throw new \InvalidArgumentException("Compost layers must be in range 0 ... " . self::MAX_COMPOST_LAYERS);
+		}
+		$this->layers = $layers;
+		return $this;
+	}
+
+	public function onInteract(Item $item, int $face, Vector3 $clickVector, ?Player $player = null, array &$returnedItems = []) : bool{
+		if($this->layers >= self::MAX_COMPOST_LAYERS){
+			$event = new ComposterEmptyEvent($this, self::MIN_COMPOST_LAYERS, [VanillaItems::BONE_MEAL()]);
+			$event->call();
+
+			if($event->isCancelled()){
+				return true;
+			}
+
+			$this->setCompostLayers($event->getCompostLayer());
+			$this->position->getWorld()->setBlock($this->position, $this);
+			$this->position->getWorld()->addSound($this->position, new ComposterEmptySound());
+
+			foreach($event->getDrops() as $drop){
+				$this->position->getWorld()->dropItem($this->position->add(0.5, 1.0, 0.5), $drop);
+			}
+			return true;
+		}
+
+		if(!$item->isNull() && $item->isCompostable()){
+			$event = new ComposterFillEvent(
+				$this, 
+				$item, 
+				$this->layers,
+				$this->layers + 1, 
+				mt_rand(1, 100) <= $item->getCompostabilityChance() && $this->layers < self::MAX_COMPOST_LAYERS - 1
+			);
```

Comment:
Instead of having a `result` field, I think it'd be better to set the `new fill layers` property to be the same as the old if the chance of adding a layer wasn't met, which would be less confusing for plugin devs using `setNewFillLayers` (although I'm doubtful that's necessary either)